File: frontend/src/shared/relation/hierarchyLabels.js
export const DEFAULT_HIERARCHY_LABELS = {
  parent: "Родительская запись",
  child: "Дочерняя запись",
  children: "Дочерние записи",
  children_genitive: "Дочерних записей",
  children_instrumental: "Дочерними записями",
};

const HIERARCHY_LABEL_KEYS = Object.keys(DEFAULT_HIERARCHY_LABELS);

function normalizeText(value) {
  return String(value ?? "").trim();
}

function looksLikeMasculineConsonant(lower) {
  if (!lower) {
    return false;
  }

  const last = lower[lower.length - 1];
  return !"аеёиоуыэюяь".includes(last);
}

/**
 * MVP Russian inflection heuristics for hierarchy terminology in Studio.
 *
 * @param {string} child
 * @param {string} [parent]
 */
export function suggestRussianHierarchyInflection(child, parent = "") {
  const singular = normalizeText(child);
  const parentLabel = normalizeText(parent);

  if (!singular) {
    return { ...DEFAULT_HIERARCHY_LABELS };
  }

  const lower = singular.toLowerCase();
  let children = singular;
  let genitive = singular;
  let instrumental = singular;

  if (lower.endsWith("ие") && singular.length > 2) {
    const stem = singular.slice(0, -2);
    children = `${stem}ия`;
    genitive = `${stem}ий`;
    instrumental = `${stem}иями`;
  } else if (lower.endsWith("ия") && singular.length > 2) {
    const stem = singular.slice(0, -2);
    children = `${stem}ии`;
    genitive = `${stem}ий`;
    instrumental = `${stem}иями`;
  } else if (lower.endsWith("ь")) {
    const stem = singular.slice(0, -1);
    children = `${stem}и`;
    genitive = `${stem}ей`;
    instrumental = `${stem}ями`;
  } else if (lower.endsWith("а")) {
    const stem = singular.slice(0, -1);
    children = `${stem}и`;
    genitive = stem;
    instrumental = `${stem}ами`;
  } else if (lower.endsWith("я")) {
    const stem = singular.slice(0, -1);
    children = `${stem}и`;
    genitive = `${stem}й`;
    instrumental = `${stem}ми`;
  } else if (looksLikeMasculineConsonant(lower)) {
    children = `${singular}ы`;
    genitive = `${singular}ов`;
    instrumental = `${singular}ами`;
  }

  return {
    parent: parentLabel || DEFAULT_HIERARCHY_LABELS.parent,
    child: singular,
    children,
    children_genitive: genitive,
    children_instrumental: instrumental,
  };
}

function readStoredLabels(settings) {
  if (!settings || typeof settings !== "object") {
    return {};
  }

  const raw = settings.hierarchy_labels;
  if (!raw || typeof raw !== "object") {
    return {};
  }

  return Object.fromEntries(
    HIERARCHY_LABEL_KEYS.map((key) => [key, normalizeText(raw[key])]).filter(
      ([, value]) => Boolean(value),
    ),
  );
}

/**
 * @param {Record<string, unknown> | null | undefined} relation
 */
export function resolveHierarchyLabelsFromRelation(relation) {
  const settings =
    relation?.settings_json && typeof relation.settings_json === "object"
      ? relation.settings_json
      : {};

  const stored = readStoredLabels(settings);
  return {
    ...DEFAULT_HIERARCHY_LABELS,
    ...stored,
  };
}

/**
 * @param {Record<string, unknown> | null | undefined} catalog
 * @param {string | null | undefined} objectTypeKey
 * @param {string | null | undefined} relationKey
 */
export function resolveHierarchyLabelsFromCatalog(catalog, objectTypeKey, relationKey) {
  const key = normalizeText(relationKey);

  if (!key) {
    return { ...DEFAULT_HIERARCHY_LABELS };
  }

  const relations = Array.isArray(catalog?.relations) ? catalog.relations : [];
  const relation = relations.find((item) => normalizeText(item?.key) === key);

  return resolveHierarchyLabelsFromRelation(relation);
}

/**
 * @param {Partial<typeof DEFAULT_HIERARCHY_LABELS> | null | undefined} labels
 */
export function resolveHierarchyLabels(labels) {
  if (!labels || typeof labels !== "object") {
    return { ...DEFAULT_HIERARCHY_LABELS };
  }

  return {
    ...DEFAULT_HIERARCHY_LABELS,
    ...Object.fromEntries(
      HIERARCHY_LABEL_KEYS.map((key) => [key, normalizeText(labels[key])]).filter(
        ([, value]) => Boolean(value),
      ),
    ),
  };
}

function toRussianChildAccusative(child) {
  const text = normalizeText(child);

  if (!text) {
    return "Дочернюю запись";
  }

  if (text.toLowerCase() === DEFAULT_HIERARCHY_LABELS.child.toLowerCase()) {
    return "Дочернюю запись";
  }

  const lower = text.toLowerCase();

  if (lower.endsWith("а")) {
    return `${text.slice(0, -1)}у`;
  }

  if (lower.endsWith("я")) {
    return `${text.slice(0, -1)}ю`;
  }

  return text;
}

/**
 * Row menu label: «Создать подзадачу» from child nominative.
 *
 * @param {string} child
 */
export function formatCreateChildMenuLabel(child) {
  const text = normalizeText(child);

  if (!text) {
    return "Создать дочернюю запись";
  }

  const accusative = toRussianChildAccusative(text);
  return `Создать ${accusative.toLowerCase()}`;
}

function sentenceCaseChildren(children) {
  const text = normalizeText(children);

  if (!text) {
    return DEFAULT_HIERARCHY_LABELS.children.toLowerCase();
  }

  return text.charAt(0).toLowerCase() + text.slice(1);
}

/**
 * @param {Partial<typeof DEFAULT_HIERARCHY_LABELS> | null | undefined} labels
 */
export function buildObjectEntityDeleteScenarioOptions(labels) {
  const resolved = resolveHierarchyLabels(labels);
  const childrenLower = sentenceCaseChildren(resolved.children);

  return [
    {
      value: "unlink_children",
      title: "Удалить только запись",
      description: `${resolved.children} сохранятся. Связи будут удалены.`,
    },
    {
      value: "with_descendants",
      title: `Удалить запись и все ${childrenLower}`,
      description: "Будет удалена вся ветка связанных записей. Действие нельзя отменить.",
    },
  ];
}

/**
 * @param {Partial<typeof DEFAULT_HIERARCHY_LABELS> | null | undefined} labels
 */
export function formatDeleteScenarioSubtitle(labels) {
  const resolved = resolveHierarchyLabels(labels);
  const childrenLower = sentenceCaseChildren(resolved.children);

  return `У записи обнаружены связанные ${childrenLower}. Выберите способ удаления.`;
}

function withInstrumentalPreposition(instrumental) {
  const text = normalizeText(instrumental);

  if (!text) {
    const fallback = DEFAULT_HIERARCHY_LABELS.children_instrumental;
    return `С ${fallback.charAt(0).toLowerCase()}${fallback.slice(1)}`;
  }

  return `С ${text.charAt(0).toLowerCase()}${text.slice(1)}`;
}

/**
 * Bulk delete scenario subtitle from hierarchy_labels.
 *
 * @param {Partial<typeof DEFAULT_HIERARCHY_LABELS> | null | undefined} labels
 */
export function formatBulkDeleteScenarioSubtitle(labels) {
  const resolved = resolveHierarchyLabels(labels);
  const childrenLower = sentenceCaseChildren(resolved.children);

  return `У части выбранных записей обнаружены связанные ${childrenLower}. Выберите способ удаления.`;
}

/**
 * @param {Partial<typeof DEFAULT_HIERARCHY_LABELS> | null | undefined} labels
 */
export function buildObjectEntityBulkDeleteScenarioOptions(labels) {
  const resolved = resolveHierarchyLabels(labels);
  const childrenLower = sentenceCaseChildren(resolved.children);

  return [
    {
      value: "unlink_children",
      title: "Удалить только выбранные записи",
      description: `${resolved.children} сохранятся. Связи с удаляемыми записями будут удалены.`,
    },
    {
      value: "with_descendants",
      title: `Удалить выбранные записи и все ${childrenLower}`,
      description: `Будут удалены все найденные ${childrenLower}. Действие нельзя отменить.`,
    },
  ];
}

/**
 * @param {{ selectedCount?: number, recordsWithChildren?: number, totalChildren?: number } | null | undefined} aggregate
 * @param {Partial<typeof DEFAULT_HIERARCHY_LABELS> | null | undefined} labels
 */
export function buildBulkDeleteStatsBadges(aggregate, labels) {
  const resolved = resolveHierarchyLabels(labels);
  const stats = aggregate || {};

  return [
    {
      label: "Выбрано",
      value: Number(stats.selectedCount) || 0,
    },
    {
      label: withInstrumentalPreposition(resolved.children_instrumental),
      value: Number(stats.recordsWithChildren) || 0,
    },
    {
      label: resolved.children_genitive,
      value: Number(stats.totalChildren) || 0,
    },
  ];
}

/**
 * @param {Partial<typeof DEFAULT_HIERARCHY_LABELS> | null | undefined} labels
 */
export function buildBulkDeleteWithDescendantsWarningItems(labels) {
  const resolved = resolveHierarchyLabels(labels);
  const childrenLower = sentenceCaseChildren(resolved.children);

  return [
    "выбранные записи",
    `все найденные ${childrenLower}`,
    "связи между ними",
  ];
}

/**
 * Presentation bundle for bulk entity delete modal.
 *
 * @param {Partial<typeof DEFAULT_HIERARCHY_LABELS> | null | undefined} labels
 * @param {{ selectedCount?: number, recordsWithChildren?: number, totalChildren?: number } | null | undefined} aggregate
 */
export function buildBulkDeleteLabels(labels, aggregate) {
  return {
    scenarioSubtitle: formatBulkDeleteScenarioSubtitle(labels),
    scenarioOptions: buildObjectEntityBulkDeleteScenarioOptions(labels),
    statsBadges: buildBulkDeleteStatsBadges(aggregate, labels),
    warningItems: buildBulkDeleteWithDescendantsWarningItems(labels),
  };
}

File: backend/app/modules/platform/shared/hierarchy_labels.py
"""Hierarchy relation terminology (labels) for universal Object Table UI."""

from __future__ import annotations

from typing import Any, TypedDict


class HierarchyLabelsDict(TypedDict):
    parent: str
    child: str
    children: str
    children_genitive: str
    children_instrumental: str


DEFAULT_HIERARCHY_LABELS: HierarchyLabelsDict = {
    "parent": "Родительская запись",
    "child": "Дочерняя запись",
    "children": "Дочерние записи",
    "children_genitive": "Дочерних записей",
    "children_instrumental": "Дочерними записями",
}

_HIERARCHY_LABEL_KEYS = tuple(DEFAULT_HIERARCHY_LABELS.keys())


def _normalize_text(value: Any) -> str:
    return str(value or "").strip()


def suggest_russian_hierarchy_inflection(child: str, parent: str = "") -> HierarchyLabelsDict:
    """MVP Russian inflection heuristics for hierarchy labels."""
    singular = _normalize_text(child)
    parent_label = _normalize_text(parent)

    if not singular:
        return dict(DEFAULT_HIERARCHY_LABELS)

    lower = singular.lower()
    children = singular
    genitive = singular
    instrumental = singular

    if lower.endswith("ие") and len(singular) > 2:
        stem = singular[:-2]
        children = f"{stem}ия"
        genitive = f"{stem}ий"
        instrumental = f"{stem}иями"
    elif lower.endswith("ия") and len(singular) > 2:
        stem = singular[:-2]
        children = f"{stem}ии"
        genitive = f"{stem}ий"
        instrumental = f"{stem}иями"
    elif lower.endswith("ь"):
        stem = singular[:-1]
        children = f"{stem}и"
        genitive = f"{stem}ей"
        instrumental = f"{stem}ями"
    elif lower.endswith("а"):
        stem = singular[:-1]
        children = f"{stem}и"
        genitive = stem
        instrumental = f"{stem}ами"
    elif lower.endswith("я"):
        stem = singular[:-1]
        children = f"{stem}и"
        genitive = f"{stem}й"
        instrumental = f"{stem}ми"
    elif _looks_like_masculine_consonant(lower):
        children = f"{singular}ы"
        genitive = f"{singular}ов"
        instrumental = f"{singular}ами"

    return {
        "parent": parent_label or DEFAULT_HIERARCHY_LABELS["parent"],
        "child": singular,
        "children": children,
        "children_genitive": genitive,
        "children_instrumental": instrumental,
    }


def _looks_like_masculine_consonant(lower: str) -> bool:
    if not lower:
        return False
    return lower[-1] not in "аеёиоуыэюяь"


def _read_stored_labels(settings: dict[str, Any] | None) -> dict[str, str]:
    if not isinstance(settings, dict):
        return {}

    raw = settings.get("hierarchy_labels")
    if not isinstance(raw, dict):
        return {}

    return {
        key: _normalize_text(raw.get(key))
        for key in _HIERARCHY_LABEL_KEYS
        if _normalize_text(raw.get(key))
    }


def resolve_hierarchy_labels_from_relation(
    relation: dict[str, Any] | None,
    *,
    parent_object_type_name: str = "",
) -> HierarchyLabelsDict:
    del parent_object_type_name

    settings = relation.get("settings_json") if isinstance(relation, dict) else {}
    if not isinstance(settings, dict):
        settings = {}

    stored = _read_stored_labels(settings)
    result = dict(DEFAULT_HIERARCHY_LABELS)
    result.update(stored)
    return result  # type: ignore[return-value]


def sanitize_hierarchy_labels_payload(raw: dict[str, Any] | None) -> dict[str, str]:
    if not isinstance(raw, dict):
        return {}

    return {
        key: _normalize_text(raw.get(key))
        for key in _HIERARCHY_LABEL_KEYS
        if _normalize_text(raw.get(key))
    }


def merge_hierarchy_settings(
    settings: dict[str, Any] | None,
    *,
    is_hierarchy: bool | None = None,
    hierarchy_labels: dict[str, Any] | None = None,
) -> dict[str, Any]:
    merged = dict(settings) if isinstance(settings, dict) else {}

    if is_hierarchy is not None:
        merged["is_hierarchy"] = bool(is_hierarchy)

    if hierarchy_labels is not None:
        sanitized = sanitize_hierarchy_labels_payload(hierarchy_labels)
        if sanitized:
            merged["hierarchy_labels"] = sanitized
        elif "hierarchy_labels" in merged:
            del merged["hierarchy_labels"]

    return merged
